text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,209 @@
+- Feature Name: ptr-to-field
+- Start Date: 2019-06-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This feature could serve as the backbone for some pointer to field syntax, and even if no syntax is made, this feature serves as a safe generic way to talk about types and their fields.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this feature is to allow safe projection through smart pointers, for example `Pin<&mut T>` to `Pin<&mut Field>`. This is a much needed feature to make `Pin<P>` more usable in safe-contexts, without the need to use unsafe to map to a field. This also can allow projection through other smart pointers like `Rc<T>`, `Arc<T>`. This feature cannot be implemented as a library effectively because it depends on the layouts of types, so it requires integration with the Rust compiler until Rust gets a stable layout (which may never happen).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+First the core traits, type, and functions that need to be added
+
+```rust
+/// Contains metadata about how to get to the field from the parent using raw pointers
+/// This is an opaque type that never needs to be stablized, and it only an implementation detail
+struct FieldDescriptor {
+    ...
+}
+
+/// The compiler should prevent user implementations for `Field`,
+/// i.e. only the compiler is allowed to implement `Field`
+/// This is to prevent people from creating fake "fields"
+trait Field {
```

Comment:
This should be `unsafe trait` to represent that unsafe code can rely on its values.

Reply:
Ojh, yes, thanks for the catch!